src/window.c

```diff
@@ -6379,9 +6379,10 @@ win_new_height(win_T *wp, int height)
 	    if (curwin->w_cursor.lnum
 		    > (curwin->w_botline - 1 - get_scrolloff_value()))
 	    {
-	        curwin->w_fraction = (height - get_scrolloff_value() - 0.0) / height
-								    * FRACTION_MULT;
+	        curwin->w_fraction = (float)(height - get_scrolloff_value())
+						     / height * FRACTION_MULT;
 	        scroll_to_fraction(wp, prev_height);
+		curwin->w_valid &= ~VALID_TOPLINE;
```

Comment:
Without this invalidation, it was possible to scroll to invalid cursor positions disobeying `'scrolloff'`. I guess a consequence of setting `w_fraction` manually.